py/compile: Add optimisation to compile OrderedDict inplace.

dpgeorge opened the pull request:
Alternative to #4259 that does everything in the compiler.

Still to optimise: eliminate need for `ROT_TWO` opcode after each key/value load.

pfalcon commented:
Thanks. I'm not versed well enough in the parser/compiler to be 100% sure, and would need to build and run this code, but what I see here is that for 1000-element OrderedDict, its elements will be stashed on the stack (which first needs to be allocated to hold those 1000 elements), then a real 1000-element OrderedDict is allocated and items are copied there from stack.

But as clearly explained in https://github.com/micropython/micropython/pull/4259, its whole idea is to be as efficient as normal dict construction, which avoids double allocations like above. And a practical usecase for OrderedDict literals introduction is `uctypes` structure definitions, which are indeed expected to be large and double-allocation being a problem.

dpgeorge commented:
>  but what I see here is that for 1000-element OrderedDict, its elements will be stashed on the stack (which first needs to be allocated to hold those 1000 elements), then a real 1000-element OrderedDict is allocated and items are copied there from stack.

Yes that is correct (same situation as tuple, list and set; although in some cases it can use the C stack for the Python stack).  It's a trade-off: in the case here the bytecode is smaller because it just loads the keys and values, then builds it all at once.  And since construction is considered to be a once off thing, it's better to optimise for small bytecode (which will stick around, but is only used once).  Also, the Python value stack for the function is reused to construct all the objects that are constructed at that time, eg multiple lists, OrderedDicts, etc.

dpgeorge commented:
I pushed a commit which modifies this to build the OrderdDict one item at a time, as the arguments are compiled.

pfalcon commented:
Ok, thanks. To compare, my original PR https://github.com/micropython/micropython/pull/4259 compiled `OrderedDict({1: 2, 3: 4})` to:

~~~ 
20 BUILD_ORDERED_MAP 2
22 LOAD_CONST_SMALL_INT 2
23 LOAD_CONST_SMALL_INT 1
24 STORE_MAP
25 LOAD_CONST_SMALL_INT 4
26 LOAD_CONST_SMALL_INT 3
27 STORE_MAP
~~~

While this PR compiled it to:
~~~
20 LOAD_NAME OrderedDict (cache=0)
24 CALL_FUNCTION n=0 nkw=0
26 LOAD_CONST_SMALL_INT 2
27 LOAD_CONST_SMALL_INT 1
28 STORE_MAP
29 LOAD_CONST_SMALL_INT 4
30 LOAD_CONST_SMALL_INT 3
31 STORE_MAP
~~~

I'd still argue that my implementation is more efficient, as it precreates the OrderedDict to the exact size, while impl in this PR will be subject to exponential relocations. But let's move on with this, please merge.

pfalcon commented:
>  It's a trade-off: in the case here the bytecode is smaller because it just loads the keys and values, then builds it all at once.

Yes, there's definitely a tradeoff, and e.g. CPython switched form "add elements one by one" to "construct entire image on stack" between 3.4 and 3.5. Then they again changed process in 3.7, to have a list of constant keys to be loaded form tuple from const table. And all that is just half-way to having a "dict template" for const dicts in const table. It would be nice to choose various options like that, but IMHO that totally doesn't scale to be written in C. That's why I started https://github.com/pfalcon/python-compiler . And default C compiler apparently should optimize for minimal RAM memory usage.

Of similar note, bytecodes like LIST_APPEND, SET_ADD exist *solely* to compile list/set comprehensions, and why comprehensions are always faster than the corresponding explicit for loops. In this regard, BUILD_ORDERED_MAP could be see as such an optimized bytecode (optimizing again for memory alloc process, not runtime performance).

dpgeorge commented:
> I'd still argue that my implementation is more efficient, as it precreates the OrderedDict to the exact size, while impl in this PR will be subject to exponential relocations

Yes.  A way to improve on that could be to allow to pass an integer as the single arg to the dict/OrderedDict constructor, which would mean "preallocate for this many entries".  This would not clash with the existing constructor use, and could even be used by the user if they really wanted to precreate a dict of a certain initial size before it's populated.  Then this PR here could also use that feature to precreate the OrderedDict of the right size, then populate it.

To be explicit, the patch to the dict constructor would be:
```diff
--- a/py/objdict.c
+++ b/py/objdict.c
@@ -80,7 +80,13 @@ STATIC void dict_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_
 }
 
 STATIC mp_obj_t dict_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
-    mp_obj_t dict_out = mp_obj_new_dict(0);
+    mp_uint_t alloc = 0;
+    if (n_args == 1 && mp_obj_is_small_int(args[0])) {
+        alloc = MP_OBJ_SMALL_INT_VALUE(args[0]);
+        --n_args;
+        ++args;
+    }
+    mp_obj_t dict_out = mp_obj_new_dict(alloc);
     mp_obj_dict_t *dict = MP_OBJ_TO_PTR(dict_out);
     dict->base.type = type;
     #if MICROPY_PY_COLLECTIONS_ORDEREDDICT
```

pfalcon commented:
> A way to improve on that could be to allow to pass an integer as the single arg to the dict/OrderedDict constructor, which would mean "preallocate for this many entries".

Yes, except that we'd like to have a consistent way to do that for any of builtin containers - list, dict, etc. If you remember, there was a talk on Go-like function in `sys`, e.g. `sys.new(list, 10)`.

Anyway, I'd suggest to not complicate this PR, and merge it as is, anything else would be optimization on top of it.

dpgeorge committed:
py/compile: Add optimisation to compile OrderedDict inplace.

This optimisation eliminates the need to create a temporary normal dict.

Thanks to @pfalcon for the initial idea and implementation.

dpgeorge commented:
Ok, I force pushed this as a single commit, cleaned up and with a #if to only include the optimisation if OrderedDict is enabled.

dpgeorge commented:
Code size change for this PR is:
```
   bare-arm:    +4 +0.006% 
minimal x86:    +0 +0.000% 
   unix x64:  +184 +0.037% 
unix nanbox:  +200 +0.045% 
      stm32:   +80 +0.022% 
     cc3200:   +80 +0.043% 
    esp8266:  +136 +0.021% 
      esp32:   +92 +0.008% 
```

dpgeorge commented:
Considering the code increase to the cc3200 port which would rather have the extra 80 bytes as RAM on the heap, probably this whole patch should be guarded by a separate config option like `MICROPY_COMP_CONST_LITERAL` (which could be used for other literals opts in the future).

pfalcon commented:
> MICROPY_COMP_CONST_LITERAL
separate config option like `MICROPY_COMP_CONST_LITERAL` (which could be used for other literals opts in the future).

Sounds good, would be nice to have that done and move on to other things.

dpgeorge commented:
Merged in 0779693c23fcc55993a06e4e0adf30a3d41f90f5, cc3200 had optimisation disabled in 47e551ba59b1bb8393b649b4a5f51f1c83d889b6
